### PR TITLE
[sfp-refactoring] Initial support for CMIS application initialization

### PR DESCRIPTION
**Pull request by ds952811**
Signed-off-by: Dante Su <dante.su@broadcom.com>

#### Why I did it

Add initial support for CMIS application initialization

#### How I did it

Update sfputil and sfpshow to support CMIS application advertisement and error detection

#### How to verify it

1. Issue 'show logging xcvrd' to check if the application initialization succeeded
2. Issue 'show interfaces transceiver eeprom --dom' to verify both the static and DOM information parsers
3. Issue 'sudo sfputil show eeprom --dom' to verify both the static and DOM information parsers

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

#### Description for the changelog

<table>
<tr>
<th>Repo</th>
<th>PR title</th>
<th>State</th>
</tr>
<tr>
<td>SONiC</td>
<td><a href='https://github.com/Azure/SONiC/pull/876'>[sfp-refactoring] Initial support for CMIS application initialization</a></td>
<td></td>
</tr>
<tr>
<td>sonic-platform-common</td>
<td><a href='https://github.com/Azure/sonic-platform-common/pull/219'>[sfp-refactoring] Initial support for CMIS application initialization</a></td>
<td></td>
</tr>
<tr>
<td>sonic-platform-daemon</td>
<td><a href='https://github.com/Azure/sonic-platform-daemons/pull/217'>[sfp-refactoring] Initial support for CMIS application initialization</a></td>
<td></td>
</tr>
<td>sonic-utilities</td>
<td><a href='https://github.com/Azure/sonic-utilities/pull/1855'>[sfp-refactoring] Initial support for CMIS application initialization</a></td>
<td></td>
</table>

- Previous command output (if the output of a command-line utility has changed)

```
admin@sonic:~$ show interfaces transceiver eeprom Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement: {1: {'ap_sel_code_id': 1, 'host_electrical_interface_id': '400GAUI-8 C2M (Annex 120E)', 'module_media_interface_id': '400GBASE-DR4 (Cl 124)', 'host_lane_count': 8, 'media_lane_count': 4, 'host_lane_assignment_options': 1, 'media_lane_assignment_options': None}, 2: {'ap_sel_code_id': 2, 'host_electrical_interface_id': '100GAUI-2 C2M (Annex 135G)', 'module_media_interface_id': '100G-FR/100GBASE-FR1 (Cl 140)', 'host_lane_count': 2, 'media_lane_count': 1, 'host_lane_assignment_options': 85, 'media_lane_assignment_options': None}}
        Connector: SN optical connector
        Encoding: N/A
        Extended Identifier: Power Class 6 (12.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2020-10-07
        Vendor Name: AVAGO
        Vendor OUI: 00-17-6a
        Vendor PN: AFCT-93DRPHZ-AZ2
        Vendor Rev: 01
        Vendor SN: FD2038FG0FY

admin@sonic:~$ sudo sfputil show eeprom -p Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement: {1: {'ap_sel_code_id': 1, 'host_electrical_interface_id': '400GAUI-8 C2M (Annex 120E)', 'module_media_interface_id': '400GBASE-DR4 (Cl 124)', 'host_lane_count': 8, 'media_lane_count': 4, 'host_lane_assignment_options': 1, 'media_lane_assignment_options': None}, 2: {'ap_sel_code_id': 2, 'host_electrical_interface_id': '100GAUI-2 C2M (Annex 135G)', 'module_media_interface_id': '100G-FR/100GBASE-FR1 (Cl 140)', 'host_lane_count': 2, 'media_lane_count': 1, 'host_lane_assignment_options': 85, 'media_lane_assignment_options': None}}
        Connector: SN optical connector
        Encoding: N/A
        Extended Identifier: Power Class 6 (12.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2020-10-07
        Vendor Name: AVAGO
        Vendor OUI: 00-17-6a
        Vendor PN: AFCT-93DRPHZ-AZ2
        Vendor Rev: 01
        Vendor SN: FD2038FG0FY
```

- New command output (if the output of a command-line utility has changed)

```
admin@sonic:~$ show interfaces transceiver eeprom Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement:
                1: 400GAUI-8 C2M (Annex 120E) | 400GBASE-DR4 (Cl 124)
                2: 100GAUI-2 C2M (Annex 135G) | 100G-FR/100GBASE-FR1 (Cl 140)
        Connector: SN optical connector
        Encoding: N/A
        Extended Identifier: Power Class 6 (12.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2020-10-07
        Vendor Name: AVAGO
        Vendor OUI: 00-17-6a
        Vendor PN: AFCT-93DRPHZ-AZ2
        Vendor Rev: 01
        Vendor SN: FD2038FG0FY

admin@sonic:~$ sudo sfputil show eeprom -p Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement:
                1: 400GAUI-8 C2M (Annex 120E) | 400GBASE-DR4 (Cl 124)
                2: 100GAUI-2 C2M (Annex 135G) | 100G-FR/100GBASE-FR1 (Cl 140)
        Connector: SN optical connector
        Encoding: N/A
        Extended Identifier: Power Class 6 (12.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2020-10-07
        Vendor Name: AVAGO
        Vendor OUI: 00-17-6a
        Vendor PN: AFCT-93DRPHZ-AZ2
        Vendor Rev: 01
        Vendor SN: FD2038FG0FY
```

### CODE REVIEW

File: sfputil/main.py [hunk 1]
```diff
@@ -273,6 +273,18 @@ def convert_sfp_info_to_output_string(sfp_info_dict):
                         output += '{}{}: {}\n'.format((indent * 2), compliance_key, spec_compliance_dict[compliance_key])
                 except ValueError as e:
                     output += '{}N/A\n'.format((indent * 2))
+        elif key == 'application_advertisement':
+            try:
```
Comment (keboliu) [inline sfputil/main.py:277]: if the CLI output has been updated, I believe you need to update the unit test accordingly.

File: sfputil/main.py [hunk 2]
```diff
@@ -864,6 +880,100 @@ def on(port_name):
     """Enable low-power mode for SFP transceiver"""
     set_lpmode(port_name, True)
 
+# 'show loopback' subcommand
+@show.command()
+@click.option('-p', '--port', metavar='<port_name>', help="Display SFP low-power mode status for port <port_name> only")
```
Comment (keboliu) [inline sfputil/main.py:885]: suggest adding a unit test to cover this new CLI

File: sfputil/main.py [hunk 3]
```diff
@@ -864,6 +880,100 @@ def on(port_name):
...
+    else:
+        if platform_sfputil.is_logical_port(port) == 0:
+            click.echo("Error: invalid port '{}'\n".format(port))
+            print_all_valid_port_values()
+            sys.exit(ERROR_INVALID_PORT)
+
+        logical_port_list = [port]
+
+    for logical_port in logical_port_list:
+        physical_port_list = logical_port_name_to_physical_port_list(logical_port)
```
Comment (keboliu) [inline sfputil/main.py:949]: in which case there will be multiple physical ports underneath one logical port?

File: sfputil/main.py (hunk 1 above)
Comment (ds952811) [reply sfputil/main.py:277]: UT is available he
[CMIS_UT_IX9.log](https://github.com/Azure/sonic-utilities/files/7293691/CMIS_UT_IX9.log)
re

File: sfputil/main.py (hunk 2 above)
Comment (ds952811) [reply sfputil/main.py:885]: Please check the UT result above

**Review by Junchao-Mellanox — DISMISSED** [review]

**Comment by prgeor** [comment]
@ds952811 could you run sonic-mgmt [tests](https://github.com/Azure/sonic-mgmt/tree/master/tests/platform_tests/sfp)

**Pull request closed without merging**